Port to TypeScript

chelkyl opened the pull request:
Changes include:
* switched from yarn to npm
* switched from rollup to tsup
* switched from ava+c8+codecov to jest
* updated dependencies excluding pako and eslint
* removed old polyfills (user can supply if needed)
* build target is ES2020
* added eslint rules
* asyncIterReader.readSize now returns an array of the read size and the read chunks instead of sometimes the read size and sometimes the read chunks for easier typing

We attempted to preserve code structure and compatibility as much as possible while adding type safety.

Thank you for creating this library!

jlarmstrongiv committed:
feat: 🎸 initial typescript support

Co-authored-by: chelkyl <chelkyl@users.noreply.github.com>

jlarmstrongiv committed:
feat: 🎸 file reorganization and start converting tests to typescript

Co-authored-by: chelkyl <chelkyl@users.noreply.github.com>

jlarmstrongiv committed:
feat: 🎸 continue converting tests to jest with typescript

Co-authored-by: chelkyl <chelkyl@users.noreply.github.com>

ikreymer committed:
package: remove node engine requirements, as its too restrictive (don't need 18.x for rollup build)
bump to 1.6.1

jlarmstrongiv committed:
ci: 🎡 upgrade to node 18 and use jest for coverage

Co-authored-by: chelkyl <chelkyl@users.noreply.github.com>

jlarmstrongiv committed:
feat: 🎸 named exports and fix invalid stream source

Co-authored-by: chelkyl <chelkyl@users.noreply.github.com>

chelkyl committed:
feat: add jest test debug config for vscode

chelkyl committed:
refactor: simplify types

chelkyl committed:
fix: should check async iterator symbol first

chelkyl committed:
fix: should use async iterator w/o calling it

chelkyl committed:
refactor: remove debug param from parse

chelkyl committed:
fix: missing period expected in test

chelkyl committed:
fix: should use payload directly

not internal buffer

chelkyl committed:
fix: jest should ignore non-test dirs

chelkyl committed:
fix: incorrect type and arg name

chelkyl committed:
test: update test data file paths

chelkyl committed:
fix: missing null in constructor call

which defaults to gzip

chelkyl committed:
docs: compressed is a string

only applies if null

chelkyl committed:
fix: swap command name with alias

so actual field name matches field name in type

chelkyl committed:
fix: correctly handle null and undef field values

chelkyl committed:
test: catch and test console warn call

so it doesn't appear in jest output

chelkyl committed:
Merge remote-tracking branch 'upstream/main'

chelkyl committed:
fix: missing type from Source

chelkyl committed:
fix: runtime type errors

due to ReadableStreamDefaultReader.read
returning Promise of ReadableStreamDefaultReadResult

also remove unnecessary types

chelkyl committed:
lint: add rules and fix findings

chelkyl committed:
refactor: compression stream types

Co-authored-by: John L. Armstrong IV <20903247+jlarmstrongiv@users.noreply.github.com>

chelkyl committed:
build: use real node version

Co-authored-by: John Armstrong <jlarmstrongiv@users.noreply.github.com>

chelkyl committed:
chore: export types

chelkyl committed:
chore: cleanup eslint rule ignores

chelkyl committed:
build: fix target and remove breaking type export

Co-authored-by: John Armstrong <jlarmstrongiv@users.noreply.github.com>

chelkyl committed:
feat: commit build

chelkyl committed:
chore: cleanup tsconfig

ikreymer commented:
@chelkyl @jlarmstrongiv wow, thank you for all this work! It was on our roadmap to convert to TS eventually, but hadn't planned on it yet. We'll try to review it quickly.
Some quick questions to start:

> switched from yarn to npm
> switched from rollup to tsup
> switched from ava+c8+codecov to jest

Just wondering about any particular reasons for these changes, especially for testing one?

jlarmstrongiv commented:
@ikreymer great news!  We really like this library and we’re glad we can contribute to it 😄 thank you for helping review these changes.  Please let us know if you have more questions:

> switched from yarn to npm

Several years ago, yarn had substantial feature and performance improvements over npm.  However, many of these changes have been added back into npm, so I no longer feel the need to add an extra step with yarn.  I also haven’t closely followed yarn after they released v2/berry and the new pnpMode, with several companies staying with yarn v1/classic or switching back to npm.  It should be an easy change to revert back to yarn if you prefer it.

> switched from rollup to tsup

Tsup is a no config bundler based on esbuild and rollup, with typescript support out of the box.  Plus, it supports both ESModules and CommonJS.  I found it to be the easiest way to switch to migrate to typescript for warcio.js and other projects. 

> switched from ava+c8+codecov to jest

By switching from ava+c8+codecov to jest, we were able to replace three tools with one more popular tool that does it all.  I have also found jest to be easier to setup for TypeScript, multiple environments (nodejs and browser), and adding polyfills.

ikreymer committed:
Merge branch 'main' of github.com:jlarmstrongiv/warcio.js into jlarmstrongiv-main

ikreymer commented on package.json:
```diff
@@ -1,67 +1,69 @@
+  "module": "dist/index.js",
+  "source": "src/index.ts",
+  "types": "dist/index.d.ts",
+  "bin": "./dist/cli.js",
```
It looks like this needs to be the .cjs version for the cli command to work, without the dynamic import

```suggestion
  "bin": "./dist/cli.cjs",
```

ikreymer commented on package.json:
```diff
@@ -1,67 +1,69 @@
+    "tsup": "^6.5.0",
+    "typescript": "^4.9.3"
+  },
+  "engines": {
```
In 1.6.1, removed the >=18.0.0 requirement, as still need to be able to bundle when running in older versions of node, and that works well.

ikreymer commented on src/index.ts:
```diff
@@ -0,0 +1,41 @@
+  WARCType,
+} from "./lib";
+
+export { main, indexCommandArgs, cdxIndexCommandArgs } from "./commands";
```
The `./commands` dir probably shouldn't be exported as part of the main export, as that breaks the browser build in webpack, as it brings in yargs. This should only be part of the cli. The `Indexer` and `CDXIndexer` can be used directly without any node dependencies.

ikreymer commented on src/lib/index.ts:
```diff
@@ -0,0 +1,38 @@
+  postToGetUrl,
+  getSurt,
+  appendRequestQuery,
+  jsonToQueryString,
```
Need to include all of the utils exports (at least for now, as they're used in other libs that use warcio.js)
```suggestion
  jsonToQueryParams,
  mfdToQueryParams,
  jsonToQueryString,
  splitChunks,
```

ikreymer commented on src/index.ts:
```diff
@@ -0,0 +1,41 @@
+  postToGetUrl,
+  getSurt,
+  appendRequestQuery,
+  jsonToQueryString,
```
Same as in lib/index.ts, need to include additional exports
```suggestion
  jsonToQueryParams,
  mfdToQueryParams,
  jsonToQueryString,
  splitChunks,
```

ikreymer commented:
Made a couple of comments, mostly adding additional exports. Also pushed (1.6.1) release which was released but forgot to push to GH (sorry!) that just disables the min node version.

With these, I was able to get it to build in wabac.js, which is one of the main libraries using this.

The main issue I'm running in now, is with the test suite in wabac.js is with the pako types, which in 1.0.x do not include the proper exports. This repros with just a node script that has:
```
import { WARCParser } from 'warcio';
```
which fails because pako doesn't export Inflate in 1.0.x.
What is the correct solution for this when used in JS modules? (I think there's other changes that prevented moving to pako 2.x for now). Ideally, the new warcio 2.0 could just be used in the same way in an ESM modules?

ikreymer committed:
2.0 release fixes:
- remove commands from main index.ts
- add utils.ts export
- pako: use 'import pako' as pako 1.0 is a cjs module, needed to work with upstream libs (eg. warcio.js)
- utils: add 'utils' entrypoint to only import utils (needed for wabac.js)
- remove extra pako reference

ikreymer committed:
Merge branch 'main' into jlarmstrongiv-main

ikreymer commented:
Added a commit that I think fixes the issues so far!
Just using `import pako from "pako"` to resolve the issues in upstream libraries, and also added a `utils` only export and removed cli commands from main export.

jlarmstrongiv commented:
@ikreymer thank you so much for reviewing and fixing 😄 glad it’s looking good!

ikreymer commented:
> @ikreymer thank you so much for reviewing and fixing 😄 glad it’s looking good!

Yep, Thanks for starting this! Just a couple of things left, for example:

> * asyncIterReader.readSize now returns an array of the read size and the read chunks instead of sometimes the read size and sometimes the read chunks for easier typing

Hm, I think this is less than desirable, but see why it was needed to make it work. 
Will see if can refactor it to just return the chunk read, and maybe do skipping differently..

ikreymer committed:
readSize / skipSize refactor:
- readSize returns chunks read, upto size
- skipSize skips reading, returns size skipped
- both use internal _readOrSkip() which returns array + size, but only used internally

ikreymer committed:
cli: fix cli running, pass process.argv if no args
cli: fix builder options, set type to 'any' for now
deps: replace hi-base32 -> base32-encode (esm module)
tests: fix tests to be fully esm, switch jest to esm mode

ikreymer committed:
remove 'process.stdout' from main build, even as default

ikreymer committed:
tests: move tests to be back under test/ dir to match existing convention

ikreymer committed:
lint fix

ikreymer commented on src/commands/args.ts:
```diff
@@ -0,0 +1,49 @@
+
+//export type IndexCommandArgs = Awaited<typeof indexCommandArgs.argv>;
+// todo: fix types?
+export type IndexCommandArgs = any;
```
@jlarmstrongiv @chelkyl Had to change the `indexCommandArgs` to actually be a function since it operates on a local yarg object (maybe there's a different way).
It all works, but not sure what the `IndexCommandArgs` `CdxIndexCommandArgs` should be..
`ReturnType<typeof indexCommandArgs>` seems close, but not quite.
Any thoughts?

ikreymer replied:
(Could also just leave it as `any` for now and come back to this later)

ikreymer replied:
Will come back to this later.

jlarmstrongiv replied:
I took a quick look at that today, but didn’t quite get it working either.  I’ll dive deeper to fix it later

ikreymer committed:
ci: add npm release script, switch ci yarn->npm

jlarmstrongiv commented:
Thank you so much for your help reviewing, fixing, and merging @ikreymer 🎉  glad to see this update land
